language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -315,3 +362,34 @@ onHead :: (a -> a) -> [a] -> [a]
 onHead f = \case
     [] -> []
     x:xs -> f x:xs
+
+onLast :: (a -> a) -> [a] -> [a]
+onLast f = \case
+  [] -> []
+  [l] -> [f l]
+  x : xs -> x : onLast f xs
+
+-- A type such as
+--   data Q = C { x: Int, y: Text }| G { z: Bool }
+-- has a DAML-LF representation like,
+--   record Q.C = { x: Int, y: String }
+--   record Q.G = { z: Bool }
+--   variant Q = C Q.C | G Q.G
+-- Use this function to get the definitions of 'Q.C' and 'Q.G' given
+-- 'Q'.
+assocDefDataTypes :: Module -> T.Text -> [((T.Text, T.Text), DefDataType)]
+assocDefDataTypes m c =
+  [((sup, sub ), def) | def <- defDataTypes m
+    , [sup, sub] <- [unTypeConName (dataTypeCon def)], sup == c]
+
+-- Extract the serialization code associated with a data type
+-- definition.
+serFromDef :: ([T.Text] -> c)
+           -> (T.Text, T.Text)
+           -> PackageId -> Module -> NM.NameMap Template
+           -> DefDataType -> c
+serFromDef f (c1, c2) curPkgId mod tpls = f . snd . fst . genDefDataType curPkgId (c1 `dot` c2) mod tpls
+
+-- Erase occurrences of string in text.
+erase :: String -> T.Text -> T.Text
```

Comment:
This type signature feels weird. We should probably do the packing at the call site. If the string we want to delete is a literal and we have `OverloadedStrings` turned on, then we don't need to do any explicit packing anyway.

Reply:
Packing at the call site obfuscates the code which is why I pulled it out. That said, going to remove `erase` anyway.

Reply:
Yes, that would do.